gusto_embedded/src/hooks/clientcredentials.ts

```diff
@@ -0,0 +1,205 @@
+      delete this.sessions[sessionKey];
+    }
+
+    return { response, error };
+  }
+
+  private async doTokenRequest(
+    credentials: Credentials,
+    scopes: string[]
+  ): Promise<Session> {
+    const formData = new URLSearchParams();
+    formData.append("grant_type", "system_access");
+    formData.append("client_id", credentials.clientID);
+    formData.append("client_secret", credentials.clientSecret);
+
+    if (scopes.length > 0) {
+      formData.append("scope", scopes.join(" "));
+    }
+
+    const tokenURL = `${this.baseURL}/oauth/token`;
+
+    const request = new Request(tokenURL, {
+      method: "POST",
+      headers: {
+        "Content-Type": "application/x-www-form-urlencoded",
+      },
+      body: formData,
+    });
+
+    const res = await this.client?.request(request);
+    if (!res) {
+      throw new Error("Failed to fetch token");
+    }
+
+    if (res.status < 200 || res.status >= 300) {
+      throw new Error("Unexpected status code");
+    }
+
+    const data = await res.json();
+    if (!data || typeof data !== "object") {
+      throw new Error("Unexpected response format");
+    }
+
+    if (data.token_type !== "Bearer") {
+      throw new Error("Unexpected token type");
+    }
+
+    let expiresAt: number | undefined;
+    if (data.expires_in) {
+      expiresAt = Date.now() + data.expires_in * 1000;
+    }
+
+    const sess: Session = {
+      credentials,
+      token: data.access_token,
+      scopes,
+    };
+
+    if (expiresAt !== undefined) {
+      sess.expiresAt = expiresAt;
+    }
+
+    return sess;
+  }
+
+  private async getCredentials(
+    source?: unknown | (() => Promise<unknown>)
+  ): Promise<Credentials | null> {
+    if (!source) {
+      return null;
+    }
+
+    let security = source;
+    if (typeof source === "function") {
+      security = await source();
+    }
+
+    // The client was passed a raw access token, no need to fetch one.
+    if (typeof security === "string") {
+      return null;
+    }
+
+    const out = parse(
+      security,
+      (val) =>
+        z
+          .lazy(() => PostV1PartnerManagedCompaniesSecurity$outboundSchema)
+          .parse(val),
+      "unexpected security type"
+    );
+
+    return {
+      clientID: out.clientId ?? "",
+      clientSecret: out.clientSecret ?? "",
+    };
+  }
+
+  private getSessionKey(credentials: Credentials): string {
+    const key = `${credentials.clientID}:${credentials.clientSecret}`;
+    return stringToBase64(key);
+  }
+
+  private hasRequiredScopes(
+    scopes: string[],
+    requiredScopes: string[]
+  ): boolean {
+    return requiredScopes.every((scope) => scopes.includes(scope));
+  }
+
+  private getScopes(requiredScopes: string[], sess?: Session): string[] {
+    return [...new Set((sess?.scopes ?? []).concat(requiredScopes))];
+  }
+
+  private hasTokenExpired(expiresAt?: number): boolean {
+    return !expiresAt || Date.now() + 60000 > expiresAt;
```

Comment:
What do you think about extracting the magic number `60000` into an explanatory constant?